### PR TITLE
Microsoft AI guidance- Migrating to Microsoft MCP package

### PR DESCRIPTION
**Pull request by souvikghosh04**
## Why make this change?

- Closes on: #3021 

Migrate DAB's MCP layer from the community SDK (`ModelContextProtocol` 0.3.0-preview.4) to Microsoft's official `Microsoft.ModelContextProtocol.HttpServer` (0.1.0-preview.25). This positions DAB for future Entra ID authentication, MISE integration, rate limiting, and first-party OpenTelemetry support.

## What is this change?

- Package swap: Replaced two community NuGet packages (`ModelContextProtocol` + `ModelContextProtocol.AspNetCore`) with single `Microsoft.ModelContextProtocol.HttpServer`. Added `EngThrive-MCP` ADO feed + package source mapping. Aligned all OpenTelemetry packages (1.9 to 1.12/1.13). Bumped `Configuration.Binder` (8.0 to 9.0).
- SDK API migration: Adopted new builder pattern — `WithListToolsHandler()` / `WithCallToolHandler()` replacing inline options delegates. Moved ServerInfo/Capabilities to defensive `PostConfigure<McpServerOptions>`.
- Breaking change fixes: Removed read-only `TextContentBlock.Type` assignments. Replaced deprecated OTel APIs (SetStatus/RecordException to Activity native methods). Removed unused using OpenTelemetry.Trace.
- Stdio fix: `McpToolRegistry.RegisterTool()` now skips same-instance re-registration via `ReferenceEquals` to prevent duplicate errors in stdio mode.
- CI pipeline: Added NuGetAuthenticate service connection for `EngThrive-MCP` feed across all pipeline YAMLs.
- EngThriveNugetFeedAccessForSqlDab is an ADO service connection (NuGet type) that authenticates CI pipelines to the EngThrive-MCP ADO feed (https://pkgs.dev.azure.com/msazure/One/_packaging/EngThrive-MCP/nuget/v3/index.json), which hosts the Microsoft.ModelContextProtocol.HttpServer package and its transitive dependencies. Will add to PR description.

## How was this tested?

- Unit Tests — 44 MCP-related tests pass (tool registry + telemetry), including 2 new idempotency tests for `ReferenceEquals` fix
- Integration Tests (manual) — HTTP and stdio transports validated end-to-end (initialize, tools/list, CRUD operations, shutdown)
- Build — 0 errors, 0 warnings. dotnet format --verify-no-changes clean.

### CODE REVIEW

File: src/Directory.Packages.props [hunk 1]
```diff
@@ -38,16 +38,15 @@
     <PackageVersion Include="Microsoft.Extensions.Caching.Memory" Version="8.0.1" />
     <PackageVersion Include="Microsoft.Extensions.Caching.Abstractions" Version="9.0.0" />
     <PackageVersion Include="Microsoft.Extensions.Primitives" Version="9.0.0" />
-    <PackageVersion Include="ModelContextProtocol" Version="0.3.0-preview.4" />
-    <PackageVersion Include="ModelContextProtocol.AspNetCore" Version="0.3.0-preview.4" />
+    <PackageVersion Include="Microsoft.ModelContextProtocol.HttpServer" Version="0.1.0-preview.25" />
     <PackageVersion Include="OpenTelemetry.Exporter.OpenTelemetryProtocol" Version="1.9.0" />
```
Comment (JerryNixon) [inline src/Directory.Packages.props]: Is it right that this is not also bumped to next version?

Comment (souvikghosh04) [reply src/Directory.Packages.props]: the OpenTelemetry.Exporter.OpenTelemetryProtocol from 1.9.0 to 1.13.0 (and OpenTelemetry.Instrumentation.Runtime from 1.9.0 to 1.12.0) to align all OTel packages.

File: .pipelines/templates/build-pipelines.yml [hunk 2]
```diff
@@ -9,6 +9,8 @@
 steps:
 - task: NuGetAuthenticate@1
   displayName: 'NuGet Authenticate'
+  inputs:
+    nuGetServiceConnections: 'EngThriveNugetFeedAccessForSqlDab'
```
Comment (anushakolan) [inline .pipelines/templates/build-pipelines.yml:13]: What is `EngThriveNugetFeedAccessForSqlDab`, it would be good to add a small description of this in the PR description.

File: src/Azure.DataApiBuilder.Mcp/Core/McpServerConfiguration.cs [hunk 3]
```diff
@@ -16,80 +17,86 @@ namespace Azure.DataApiBuilder.Mcp.Core
...
+                    arguments?.Dispose();
+                }
             })
             .WithHttpTransport();
 
+            // Configure underlying MCP server options defensively to avoid overwriting any defaults
+            services.PostConfigure<McpServerOptions>(options =>
+            {
+                options.ServerInfo ??= new() { Name = McpProtocolDefaults.MCP_SERVER_NAME, Version = McpProtocolDefaults.MCP_SERVER_VERSION };
+                options.ServerInfo.Name = McpProtocolDefaults.MCP_SERVER_NAME;
```
Comment (anushakolan) [inline src/Azure.DataApiBuilder.Mcp/Core/McpServerConfiguration.cs]: Why are we explicitly setting `ServerInfor.Name` and `ServerInfo.Version`, in line 94 and 95 again, when we already passed it to the ctor in 93.

File: src/Azure.DataApiBuilder.Mcp/Core/McpToolRegistry.cs [hunk 4]
```diff
@@ -37,6 +37,14 @@ public void RegisterTool(IMcpTool tool)
             // Check for duplicate tool names (case-insensitive)
             if (_tools.TryGetValue(toolName, out IMcpTool? existingTool))
             {
+                // If the same tool instance is already registered, skip silently.
+                // This can happen when both McpToolRegistryInitializer (hosted service)
+                // and McpStdioHelper register tools during stdio mode startup.
+                if (ReferenceEquals(existingTool, tool))
+                {
```
Comment (anushakolan) [inline src/Azure.DataApiBuilder.Mcp/Core/McpToolRegistry.cs:44]: nit: We can add a debug log on why are we returning.

File: src/Azure.DataApiBuilder.Mcp/Utils/McpResponseBuilder.cs [hunk 5]
```diff
@@ -34,7 +34,7 @@ public static CallToolResult BuildSuccessResult(
             {
                 Content = new List<ContentBlock>
                 {
-                    new TextContentBlock { Type = "text", Text = output }
+                    new TextContentBlock { Text = output }
```
Comment (anushakolan) [inline src/Azure.DataApiBuilder.Mcp/Utils/McpResponseBuilder.cs:37]: Why is the type being removed?

File: src/Azure.DataApiBuilder.Mcp/Azure.DataApiBuilder.Mcp.csproj [hunk 6]
```diff
@@ -4,15 +4,16 @@
...
   </PropertyGroup>
 
   <ItemGroup>
     <InternalsVisibleTo Include="Azure.DataApiBuilder.Service.Tests" />
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="ModelContextProtocol" />
-    <PackageReference Include="ModelContextProtocol.AspNetCore" />
+    <PackageReference Include="Microsoft.ModelContextProtocol.HttpServer" />
```
Comment (anushakolan) [inline src/Azure.DataApiBuilder.Mcp/Azure.DataApiBuilder.Mcp.csproj]: We are only adding `Microsoft.ModelContextProtocol.HttpServer` and not `Microsoft.ModelContextProtocol`, and removing `ModelContextProtocol`. However, we are still using `ModelContextProtocol` in the code, how will this not be throwing `package not found` errors.

File: Nuget.config [hunk 7]
```diff
@@ -3,7 +3,26 @@
...
+    <packageSource key="data_api_builder_build_packages">
+      <package pattern="*" />
+    </packageSource>
+    <!-- CI (Azure Pipelines NuGetAuthenticate) renames the source with a "feed-" prefix at runtime -->
+    <packageSource key="feed-data_api_builder_build_packages">
+      <package pattern="*" />
+    </packageSource>
+    <packageSource key="EngThrive-MCP">
+      <package pattern="Microsoft.ModelContextProtocol.*" />
+      <package pattern="ModelContextProtocol" />
```
Comment (anushakolan) [inline Nuget.config:18]: 1. Why are we having both `Microsoft.ModelContextProtocol` and `ModelContextProtocol`.
2. Which one is being used by the code?

File: src/Service.Tests/Mcp/McpToolRegistryTests.cs [hunk 8]
```diff
@@ -141,6 +141,50 @@ public void RegisterTool_WithDifferentCasing_ThrowsException()
             Assert.AreEqual(DataApiBuilderException.SubStatusCodes.ErrorInInitialization, exception.SubStatusCode);
         }
 
+        /// <summary>
+        /// Test that registering the same tool instance twice is silently ignored (idempotent).
+        /// This supports stdio mode where both McpToolRegistryInitializer and McpStdioHelper may register the same tools.
+        /// </summary>
+        [TestMethod]
```
Comment (anushakolan) [inline src/Service.Tests/Mcp/McpToolRegistryTests.cs:148]: What happens when same instance is registered under different names?

**Review by anushakolan — COMMENTED** [review]
Added some comments, which are mostly questions to understand the context better. Will review again once the questions are answered.

File: Nuget.config [hunk 9]
```diff
@@ -3,7 +3,26 @@
...
   </packageSources>
+  <packageSourceMapping>
+    <packageSource key="data_api_builder_build_packages">
+      <package pattern="*" />
+    </packageSource>
+    <!-- CI (Azure Pipelines NuGetAuthenticate) renames the source with a "feed-" prefix at runtime -->
+    <packageSource key="feed-data_api_builder_build_packages">
+      <package pattern="*" />
+    </packageSource>
+    <packageSource key="EngThrive-MCP">
```
Comment (Aniruddh25) [inline Nuget.config:16]: We need to add a similar key to our internal Official Pipeline.

File: src/Azure.DataApiBuilder.Mcp/Core/McpToolRegistry.cs (hunk 4 above)
Comment (souvikghosh04) [reply src/Azure.DataApiBuilder.Mcp/Core/McpToolRegistry.cs:44]: agree, this would be useful. however, currently there is no `ILogger` dependency and adding one would change the constructor signature, DI registration, and all test instantiations. since this is a rare edge case (only stdio mode startup), I would prefer to add logging as part of a broader MCP logging pass and not in this PR.

File: src/Azure.DataApiBuilder.Mcp/Utils/McpResponseBuilder.cs (hunk 5 above)
Comment (souvikghosh04) [reply src/Azure.DataApiBuilder.Mcp/Utils/McpResponseBuilder.cs:37]: In the new SDK version has `TextContentBlock.Type` changed to a read-only property that auto-defaults to "text". Attempting to set it causes a compile error. The behavior is identical- the type is still "text", just set by the SDK internally.

File: Nuget.config (hunk 7 above)
Comment (souvikghosh04) [reply Nuget.config:18]: `Microsoft.ModelContextProtocol.HttpServer` is the Microsoft wrapper package that DAB directly references. It has a transitive dependency on the community `ModelContextProtocol` and `ModelContextProtocol.*` packages (the underlying SDK it wraps). Both are hosted on the `EngThrive-MCP` ADO feed, so both need patterns in the source mapping. Our code only directly references `Microsoft.ModelContextProtocol.HttpServer` and the others are pulled in automatically by NuGet.

File: src/Service.Tests/Mcp/McpToolRegistryTests.cs (hunk 8 above)
Comment (souvikghosh04) [reply src/Service.Tests/Mcp/McpToolRegistryTests.cs:148]: This can't happen with the current design. the tool name is derived from `tool.GetToolMetadata().Name`, not passed as a parameter to `RegisterTool()`. The same instance always returns the same name, so ReferenceEquals correctly catches the duplicate. If a hypothetical tool returned different names on different calls, it would register under the second name (separate dictionary entry) but this would indicate a bug in the tool implementation, not in the registry.

File: Nuget.config (hunk 9 above)
Comment (souvikghosh04) [reply Nuget.config:16]: agreed, will coordinate to add the `EngThriveNugetFeedAccessForSqlDab` service connection and nuGet ServiceConnections input to the internal official pipeline as a follow-up. will sync up with you sometime.

File: src/Azure.DataApiBuilder.Mcp/Core/McpServerConfiguration.cs (hunk 3 above)
Comment (souvikghosh04) [reply src/Azure.DataApiBuilder.Mcp/Core/McpServerConfiguration.cs]: Fixed
